### PR TITLE
Provide a way to resolve pipeline context when no query is performed

### PR DESCRIPTION
**Pull request by olamothe**
The pipeline context would normally get resolved from the last query (when we are doing the call for the query suggest).

In some cases, we want to be able to provide the context without having performed any query.

This adds a logic in `SearchInterface` to `getQueryContext()`.

The search interface will return the last query context, if available.

If not available, instead it will try to resolve any `PipelineContext` component, and query them to retrieve the context.

This also adds a couple of utility method on the `PipelineContext` component to set and get the context.

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/SearchInterface/SearchInterface.ts [hunk 1]
```diff
@@ -513,6 +516,43 @@ export class SearchInterface extends RootComponent implements IComponentBindings
   }
 
   /**
+   * Get the Query context for the current Search Interface.
+   * 
+   * If the search interface has performed at least one query, it will try to resolve the context from the last query sent to the Coveo Search API.
```
Comment (francoislg) [inline src/ui/SearchInterface/SearchInterface.ts]: Hm, are you sure you want to return the last query?

Say I am dynamically changing context values based on events in the page (like "Did the user add a rating to a product), this might not trigger a query but the context value is now different and the old one will still be used for query suggestions.

File: src/ui/PipelineContext/PipelineContext.ts [hunk 2]
```diff
@@ -44,63 +46,97 @@ export interface IPipelineContextOptions {}
...
       const contextValues = [];
-      _.each(this.content[key], value => {
+      _.each(this.contextContent[key], value => {
         contextValues.push(this.getModifiedData(value));
       });
       return contextValues;
     } else if (_.isString(contextValue)) {
       return this.getModifiedData(contextValue);
     }
+    return undefined;
```
Comment (samisayegh) [inline src/ui/PipelineContext/PipelineContext.ts]: Is there a reason for choosing `undefined` over `null` or an empty string? With `undefined` it is not clear what happened if debugging. e.g. did we forget to execute `getContextValue`? or did it fail mid-execution?

File: src/ui/PipelineContext/PipelineContext.ts [hunk 3]
```diff
@@ -110,12 +146,38 @@ export class PipelineContext extends Component {
...
+     }
+
+     This means that in those case, we would try to access Coveo.context.productValueFromSalesforce (which would in theory be a "real" product value from salesforce, and not a placeholder/variable)
+    */
     return value.replace(/\{\!([^\}]+)\}/g, (all: string, contextKey: string) => {
-      if (Coveo.context != null && contextKey in Coveo.context) {
-        return Coveo.context[contextKey];
-      } else if (contextKey == PipelineContext.CURRENT_URL) {
+      const trimmedKey = contextKey.trim();
+      if (Coveo.context != null && trimmedKey in Coveo.context) {
```
Comment (samisayegh) [inline src/ui/PipelineContext/PipelineContext.ts]: Rather than checking if `Coveo.context` is not not defined, we could check if it is defined.

File: src/ui/SearchInterface/SearchInterface.ts [hunk 4]
```diff
@@ -513,6 +516,43 @@ export class SearchInterface extends RootComponent implements IComponentBindings
...
+   * 
+   * If there are multiple {@link PipelineContext} component available, it will merge all context values together.
+   * 
+   * **Note:**
+   * It is not recommended to have multiple PipelineContext component, especially if you have duplicate context key in each one.
+   * 
+   * If no context is found, returns `undefined`
+   */
+  public getQueryContext(): Context {
+    let ret: Context = undefined;
```
Comment (samisayegh) [inline src/ui/SearchInterface/SearchInterface.ts]: No need for `= undefined`

File: test/ui/PipelineContextTest.ts [hunk 5]
```diff
@@ -54,18 +54,75 @@ export function PipelineContextText() {
         it('should return the context value', () => {
           expect(test.cmp.getContextValue('qwerty')).toBe('azerty');
         });
+
+        it('should support setting a new context', () => {
+          test.cmp.setContext({ abc: 'def', hij: ['klm', 'nop'] });
+          expect(test.cmp.getContextKeys()).toEqual(jasmine.arrayContaining(['abc', 'hij']));
+        });
+
+        it('should support setting a single context value', () => {
```
Comment (samisayegh) [inline test/ui/PipelineContextTest.ts]: I think the test description should be tweaked to include the array case. Maybe something like:
`should support setting a single context value and an array of context values`.

Alternative may be to break the test into two.

File: test/ui/PipelineContextTest.ts [hunk 6]
```diff
@@ -54,18 +54,75 @@ export function PipelineContextText() {
...
+        });
+
+        describe('with a global Coveo.context variable', () => {
+          beforeEach(() => {
+            Coveo['context'] = {};
+            Coveo['context']['productName'] = 'ACME 2000 ULTIMATE EDITION';
+            Coveo['context']['userRole'] = 'SYS ADMIN';
+          });
+
+          afterEach(() => {
```
Comment (samisayegh) [inline test/ui/PipelineContextTest.ts]: Since the `beforeEach` is setting `Coveo['context']` to an empty object, is the `afterEach` necessary?

File: src/ui/PipelineContext/PipelineContext.ts [hunk 7]
```diff
@@ -44,63 +46,97 @@ export interface IPipelineContextOptions {}
  * Using this component as opposed to JavaScript code means you will be able to leverage the interface editor.
  *
  * Regardless of if you use this component or JavaScript to add context, both will add the needed data in the [Query.Context]{@link IQuery.context} parameter.
+ * 
+ * **Note:**
+ * 
+ * Using this component also ensures that the framework will be able to properly determine the context in all corner cases, including when using a Standalone Search box ({@link initSearchbox}) displaying query suggestions.
+ * 
+ * If this component is not used in those cases, the context won't be able to be resolved and leveraged properly in the [Query Pipeline](http://www.coveo.com/go?dest=cloudhelp&lcid=9&context=108).
```
Comment (fbeaudoincoveo) [inline src/ui/PipelineContext/PipelineContext.ts]: I'd rephrase as follows:

> In most cases, if you do not use this component, the context will not be resolved and leveraged properly in the query pipeline (see [What Is a Query Pipeline?](http://www.coveo.com/go?dest=cloudhelp&lcid=9&context=252)).

Note that I also changed the link to an introductory topic on query pipelines.

File: src/ui/PipelineContext/PipelineContext.ts [hunk 8]
```diff
@@ -44,63 +46,97 @@ export interface IPipelineContextOptions {}
  * Using this component as opposed to JavaScript code means you will be able to leverage the interface editor.
  *
  * Regardless of if you use this component or JavaScript to add context, both will add the needed data in the [Query.Context]{@link IQuery.context} parameter.
+ * 
+ * **Note:**
+ * 
+ * Using this component also ensures that the framework will be able to properly determine the context in all corner cases, including when using a Standalone Search box ({@link initSearchbox}) displaying query suggestions.
```
Comment (fbeaudoincoveo) [inline src/ui/PipelineContext/PipelineContext.ts]: I'd slightly rephrase this:

> This component also ensures that the framework properly determines the context in all corner cases, including when a standalone search box ([`initSearchbox`]{@link initSearchbox}) is displaying query suggestions.

File: src/ui/PipelineContext/PipelineContext.ts [hunk 9]
```diff
@@ -44,63 +46,97 @@ export interface IPipelineContextOptions {}
...
+      const contextParsed = this.tryParseContextFromString(newContext);
+      this.contextContent = contextParsed;
+    } else {
+      this.contextContent = newContext;
     }
-    this.bind.onRootElement(QueryEvents.buildingQuery, (args: IBuildingQueryEventArgs) => this.handleBuildingQuery(args));
+  }
+
+  /**
+   * Return the current context
```
Comment (fbeaudoincoveo) [inline src/ui/PipelineContext/PipelineContext.ts]: Suggestion:

> Returns the current context.

File: src/ui/PipelineContext/PipelineContext.ts [hunk 10]
```diff
@@ -44,63 +46,97 @@ export interface IPipelineContextOptions {}
...
+  public getContext(): Context {
+    const keys = this.getContextKeys();
+    return _.object(keys, _.map(keys, key => this.getContextValue(key)));
+  }
+
+  /**
+   * Set a value for a context, replacing the old value if it was already available.
+   * @param contextKey
+   * @param contextValue 
+   */
```
Comment (fbeaudoincoveo) [inline src/ui/PipelineContext/PipelineContext.ts]: Suggestion:

> Sets a value for a context key, replacing the old value if applicable.

File: src/ui/PipelineContext/PipelineContext.ts (hunk 9 above)
Comment (fbeaudoincoveo) [reply src/ui/PipelineContext/PipelineContext.ts]: I've opened a doc issue to review and uniformize method documentation (see DOC-1715).

File: src/ui/SearchInterface/SearchInterface.ts [hunk 11]
```diff
@@ -513,6 +516,43 @@ export class SearchInterface extends RootComponent implements IComponentBindings
   }
 
   /**
+   * Get the Query context for the current Search Interface.
+   * 
+   * If the search interface has performed at least one query, it will try to resolve the context from the last query sent to the Coveo Search API.
+   * 
+   * If the search interface has not performed a query yet, it will try to resolve the context from any avaiable {@link PipelineContext} component.
+   * 
+   * If there are multiple {@link PipelineContext} component available, it will merge all context values together.
```
Comment (fbeaudoincoveo) [inline src/ui/SearchInterface/SearchInterface.ts]: Suggestion:

> If multiple `PipelineContext` components are available, it will merge all context values together.

File: src/ui/SearchInterface/SearchInterface.ts [hunk 12]
```diff
@@ -513,6 +516,43 @@ export class SearchInterface extends RootComponent implements IComponentBindings
...
+   * 
+   * **Note:**
+   * It is not recommended to have multiple PipelineContext component, especially if you have duplicate context key in each one.
```
Comment (fbeaudoincoveo) [inline src/ui/SearchInterface/SearchInterface.ts]: Suggestion:

> Having multiple `PipelineContext` components in the same search interface is not recommended, especially if some context keys are repeated across those components.

File: src/ui/SearchInterface/SearchInterface.ts [hunk 13]
```diff
@@ -513,6 +516,43 @@ export class SearchInterface extends RootComponent implements IComponentBindings
   }
 
   /**
+   * Get the Query context for the current Search Interface.
```
Comment (fbeaudoincoveo) [inline src/ui/SearchInterface/SearchInterface.ts]: Suggestion:

> Gets the query context for the current search interface.
